libs/dal/src/types/mongoose.ts

```diff
@@ -0,0 +1,87 @@
+import { QuerySelector } from 'mongoose';
+import { ExtractDot, NestedDotKeys, DeepKeys } from './helpers';
+
+/**
+ * MongoDB Dot Notation: https://www.mongodb.com/docs/manual/core/document/#dot-notation
+ *
+ * Dot Notation generic with Arrays:
+ * https://stackoverflow.com/questions/76546335/dot-notation-key-extraction-for-deeply-nested-objects-arrays-and-arrays-of-obje
+ *
+ * Dot Notation generic WITHOUT Arrays:
+ * https://stackoverflow.com/questions/58434389/typescript-deep-keyof-of-a-nested-object/68404823#68404823
+ */
+
+/**
+ * Modified from original Mongoose typings to introduce stricter typing
+ * for the `FilterQuery` generic type
+ *
+ * @see https://github.com/Automattic/mongoose/blob/582156858db3ca7fbaa8950dc997e0d9e8117b21/types/query.d.ts#L4
+ */
+type ApplyBasicQueryCasting<T> = T | T[];
+type Condition<T> = ApplyBasicQueryCasting<T> | QuerySelector<ApplyBasicQueryCasting<T>>;
+
+/**
+ * Modified from original Mongoose typings to introduce stricter typing
+ * for the `FilterQuery` generic type
+ *
+ * @see https://github.com/Automattic/mongoose/blob/582156858db3ca7fbaa8950dc997e0d9e8117b21/types/query.d.ts#L70
+ */
+type RootQuerySelector<T> = {
+  /** @see https://www.mongodb.com/docs/manual/reference/operator/query/and/#op._S_and */
+  $and?: Array<FilterQuery<T>>;
+  /** @see https://www.mongodb.com/docs/manual/reference/operator/query/nor/#op._S_nor */
+  $nor?: Array<FilterQuery<T>>;
+  /** @see https://www.mongodb.com/docs/manual/reference/operator/query/or/#op._S_or */
+  $or?: Array<FilterQuery<T>>;
+  /** @see https://www.mongodb.com/docs/manual/reference/operator/query/text */
+  $text?: {
+    $search: string;
+    $language?: string;
+    $caseSensitive?: boolean;
+    $diacriticSensitive?: boolean;
+  };
+  /** @see https://www.mongodb.com/docs/manual/reference/operator/query/where/#op._S_where */
+  $where?: string;
+  /** @see https://www.mongodb.com/docs/manual/reference/operator/query/comment/#op._S_comment */
+  $comment?: string;
+} & Partial<{
+  [Key in DeepKeys<T>]: ExtractDot<T, Key> | ExtractDot<T, Key>[];
```

Comment:
We are using the helper `DeepKeys` and `ExtractDot` generics here to introduce stronger typings on the `RootQuerySelect`. This allows us to strongly type queries targeting nested attributes through Mongo's Dot Notation, as seen in `"partnerConfigurations.configurationId"?: string | string[] | undefined` below:
<img width="760" alt="image" src="https://github.com/novuhq/novu/assets/32132657/8eedb8d9-0abb-4a91-ac54-8f2183f83ffb">

There is significant work required to update the codebase to align on consistent use of these typings, because `mongoose` is very flexible by design. We will most likely want to decide on a pattern to keep the typings as simple as possible.

Other notation such as `'partnerConfigurations.$': 1` is not currently covered, and we should discuss if we'd like to include this pattern as a valid typing.

Reply:
These generic typings for `mongoose` are not presently in use, and are introduced in this PR only for discussion.